SeQuant/core/export/python_einsum.hpp

```diff
@@ -685,6 +609,176 @@ class PythonEinsumGenerator : public Generator<Context> {
   }
 };
 
+/// Generator for NumPy einsum
+class NumPyEinsumGenerator
```

Comment:
All generators should be templated based on their context type. This is needed in order to make inheritance easier. The reason being that we want don't want to require the correct context class to be used with the generator (see also https://stackoverflow.com/q/18671062).

That is, we don't want to allow `NumPyEinsumGenerator` to be fed a `PythonEinsumGeneratorContext` (which could also be the PyTorch one). This is already given thanks to CRTP on the base class. However, it would be nice to just keep the option of inheriting from `NumPyEinsumGenerator` as well (which requires it to have a CRTP for the context class).

Reply:
For "final" generators the context<->generator is 1<->1, removing ability to change context limits customization but  until we run into the need to override the default I'll leave as is.

Reply:
Yes but the problem is that handling of classes vs. template classes can be different in certain contexts. Hence, I would strongly suggest we make it a template as all other generator classes to ensure we don't have to refactor call-sides should we ever encounter the need.